src/software/backend/input/network/network_filter.cpp

```diff
@@ -0,0 +1,392 @@
+#include "software/backend/input/network/network_filter.h"
+
+#include "shared/constants.h"
+#include "software/constants.h"
+#include "software/parameter/dynamic_parameters.h"
+#include "software/proto/messages_robocup_ssl_detection.pb.h"
+#include "software/proto/messages_robocup_ssl_geometry.pb.h"
+#include "software/sensor_fusion/refbox_data.h"
+
+// We can initialize the field_state with all zeroes here because this state will never
+// be accessed by an external observer to this class. the getFieldData must be called to
+// get any field data which will update the state with the given protobuf data
+NetworkFilter::NetworkFilter()
+    : field_state(0, 0, 0, 0, 0, 0, 0, Timestamp::fromSeconds(0)),
+      ball_state(Point(), Vector(), Timestamp::fromSeconds(0)),
+      friendly_team_state(Duration::fromMilliseconds(
+          Util::Constants::ROBOT_DEBOUNCE_DURATION_MILLISECONDS)),
+      enemy_team_state(Duration::fromMilliseconds(
+          Util::Constants::ROBOT_DEBOUNCE_DURATION_MILLISECONDS)),
+      ball_filter(BallFilter::DEFAULT_MIN_BUFFER_SIZE,
+                  BallFilter::DEFAULT_MAX_BUFFER_SIZE),
+      friendly_team_filter(),
+      enemy_team_filter()
+{
+}
+
+Field NetworkFilter::getFieldData(const SSL_GeometryData &geometry_packet)
+{
+    if (geometry_packet.has_field())
+    {
+        SSL_GeometryFieldSize field_data = geometry_packet.field();
+        Field field                      = createFieldFromPacketGeometry(field_data);
+
+        field_state = field;
+    }
+
+    return field_state;
+}
+
+Field NetworkFilter::createFieldFromPacketGeometry(
+    const SSL_GeometryFieldSize &packet_geometry) const
+{
+    // We can't guarantee the order that any geometry elements are passed to us in, so
+    // We map the name of each line/arc to the actual object so we can refer to them
+    // consistently
```

Comment:
Should we be defensive here and actually check we get the geometry we need before we do all the lookups below?

Reply:
Does the proto make guarantees about having those names? What field would we return if the proto didn't have one of the lines we rely on?

Reply:
If we don't get the data we're expecting, I don't think there's a great recovery option. It's just a list of strings so there's no strict guarantee from proto itself.

Reply:
Ok, can we just check, log if we're missing any fields, and drop the packet if so?

Reply:
I think there is the "guarantee" that we get all the fields at the same time (ie. don't just get some lines in one packet, and the arcs in another), but that is not guaranteed by _proto_ itself.

What I mean by this is we should never detect a missing field just _once_ and then have all future packets be fine. If something is missing that we're expecting, it won't just appear magically in future packets. For this reason I would argue for a harder fault than just logging and dropping the packet. We should indicate a fundamental issue with the code / behavior that is irrecoverable by at the very least throwing a description exception.

Reply:
Hmm, I don't think I would reasonably expect an exception here. The program is working as expected, we are just missing some information. But the fact that the proto is fully populated really isn't guaranteed.

Reply:
Yes it is working as expected, but there is still a fairly significant flaw / violation of some assumptions. I don't expect to encounter this issue often, so at the end of the day am fairly indifferent. My main concern would just be it takes people longer to figure out what the issue is if it's just logged, but not a strong argument. Fine with logging.

Reply:
Logging is ok with me but I agree with Mat that some fundamental assumptions are being violated

Reply:
changed